javaparser-metamodel-generator/src/main/java/com/github/javaparser/generator/metamodel/MetaModelGenerator.java

```diff
@@ -0,0 +1,269 @@
+        initializeNodeMetaModelsStatements.clear();
+        initializeFieldMetaModelsStatements.clear();
+
+        for (Class<?> c : ALL_MODEL_CLASSES) {
+            String className = metaModelName(c);
+            String fieldName = decapitalize(className);
+            mmClass.getFieldByName(fieldName).ifPresent(Node::remove);
+            FieldDeclaration f = mmClass.addField(className, fieldName, PUBLIC, FINAL);
+
+            Class<?> superclass = c.getSuperclass();
+            final String superClassMetaModel = optionalOf(decapitalize(metaModelName(superclass)), isNode(superclass));
+
+            f.getVariable(0).setInitializer(parseExpression(f("new %s(this, %s)", className, superClassMetaModel)));
+            initializeNodeMetaModelsStatements.add(parseStatement(f("nodeMetaModels.add(%s);", fieldName)));
+
+            CompilationUnit classMetaModelJavaFile = new CompilationUnit(METAMODEL_PACKAGE);
+            classMetaModelJavaFile.addImport("java.util.Optional");
+            sourceRoot.add(METAMODEL_PACKAGE, className + ".java", classMetaModelJavaFile);
+            ClassOrInterfaceDeclaration classMetaModelClass = classMetaModelJavaFile.addClass(className, PUBLIC);
+            classMetaModelClass.addExtendedType(new ClassOrInterfaceType(NODE_META_MODEL));
+
+            AstTypeAnalysis typeAnalysis = new AstTypeAnalysis(c);
+
+            ConstructorDeclaration classMMConstructor = classMetaModelClass
+                    .addConstructor()
+                    .addParameter("JavaParserMetaModel", "parent")
+                    .addParameter("Optional<" + NODE_META_MODEL + ">", "super" + NODE_META_MODEL);
+            classMMConstructor
+                    .getBody()
+                    .addStatement(parseExplicitConstructorInvocationStmt(f("super(super%s, parent, %s.class, \"%s\", \"%s\", %s, %s);",
+                            NODE_META_MODEL,
+                            c.getName(),
+                            c.getSimpleName(),
+                            c.getPackage().getName(),
+                            typeAnalysis.isAbstract,
+                            typeAnalysis.isSelfType)));
+
+            generateFieldMetaModels(c, classMetaModelClass, fieldName, initializeFieldMetaModelsStatements);
+        }
+
+        initializeNodeMetaModelsStatements.sort(Comparator.comparing(Node::toString));
+    }
+
+    private boolean isNode(Class<?> c) {
+        return Node.class.isAssignableFrom(c);
+    }
+
+    private void generateFieldMetaModels(Class<?> c, ClassOrInterfaceDeclaration classMetaModelClass, String classMetaModelFieldName, NodeList<Statement> initializeFieldMetaModelsStatements) throws NoSuchMethodException {
+        List<Field> fields = new ArrayList<>(Arrays.asList(c.getDeclaredFields()));
+        fields.sort(Comparator.comparing(Field::getName));
+        for (Field field : fields) {
```

Comment:
same here 30 LOC maybe you would like fields.forEach(field -> newMethod(field)) but I'm not quite sure how expensive it would be to extract parts of this methods...